array_api_compat/common/_helpers.py

```diff
@@ -801,6 +801,174 @@ def size(x):
         return None
     return math.prod(x.shape)
 
+def is_writeable_array(x):
+    """
+    Return False if x.__setitem__ is expected to raise; True otherwise
+    """
+    if is_numpy_array(x):
+        return x.flags.writeable
+    if is_jax_array(x):
+        return False
+    return True
+
+_undef = object()
+
+def at(x, idx=_undef, /):
+    """
+    Update operations for read-only arrays.
+
+    This implements ``jax.numpy.ndarray.at`` for all backends.
+    Writeable arrays may be updated in place; you should not rely on it.
+
+    Keyword arguments (e.g. ``indices_are_sorted``) are passed to JAX and are
+    quietly ignored for backends that don't support them.
+
+    Examples
+    --------
+    Given either of these equivalent expressions::
+
+      x = at(x)[1].add(2)
+      x = at(x, 1).add(2)
+
+    If x is a JAX array, they are the same as::
+
+      x = x.at[1].add(x)
+
+    If x is a read-only numpy array, they are the same as::
+
+      x = x.copy()
+      x[1] += 2
+
+    Otherwise, they are the same as::
+
+      x[1] += 2
+
+    Warning
+    -------
+    You should always immediately overwrite the parameter array::
+
+        x = at(x, 0).set(2)
+
+    The anti-pattern below must be avoided, as it will result in different behaviour
+    on read-only versus writeable arrays:
+
+        x = xp.asarray([0, 0, 0])
+        y = at(x, 0).set(2)
+        z = at(x, 1).set(3)
+
+    In the above example, y == [2, 0, 0] and z == [0, 3, 0] when x is read-only,
+    whereas y == z == [2, 3, 0] when x is writeable!
+
+    See Also
+    --------
+    https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html
+    """
+    if is_jax_array(x):
+        return x.at
+    if is_numpy_array(x) and not x.flags.writeable:
+        x = x.copy()
+    return _InPlaceAt(x, idx)
```

Comment:
So if I understand correctly, this implementation works for:
- JAX
- NumPy
- Mutable arrays (arrays for which the in-place updates work)

It doesn't work for immutable arrays coming from other libraries, for the reasons discussed in https://github.com/data-apis/array-api/issues/845.

Reply:
It works for all arrays explicitly listed in array-api-compat, including numpy subclasses.
It won't work for immutable arrays from other, unknown libraries - in fact, they won't be even recognized as immutable. Without a standardized way for `is_immutable_array` to detect such a use case (`__array_writeable__` interface?) I would not know how to implement it.

Reply:
Would attempting to mutate the array as a fallback work, or could that lead to undesired side-effects?

```py
# Something along these lines, but more robust
a = x[0]
try:
    x[0] = 0
    x[0] = a
    mutable = True
except:
    mutable = False
```

Reply:
I think that reimplementing `at[]` without the ability to update and without specific knowledge of the library at hand would be extraordinarily inefficient.

Consider `at(x, 0).add(1)`. Implementing it on top of `where` would be extremely expensive; an alternative would be to break the array down and rebuild it with `concatenate` and `stack`, which would be very complicated and probably fragile.

The snippet of code you just wrote would perform very poorly in many cases. Consider:
- a library like dask: the `x[0] = a` line adds extra labour into the graph, which is very likely nontrivial
- any library where the memory transparently moves from device to host and vice versa would likely cause either a page or the whole array to move from device to host (and vice versa).

A somewhat slightly more reasonable alternative would be to

1. blindly try the update
2. if the update fails, try if by any chance the library has a `at[]` method with exactly the same API as JAX

At the moment we're only doing (1).
My personal opinion would be to explicitly cater for these libraries if and when they crop up.
Is this whole discussion hypothetical, or do we know of a specific read-only library other than JAX?

Reply:
I just realised that sparse.COO does not support updates. Has there been any discussion about it?